accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+### Overriding Access
+
+If two or more extensions apply to the same member access, or if a member of the receiver type takes precedence over an extension method, or if the extension is imported with a prefix, then it is possible to force an extension member invocation:
+
+```dart
+MyList(object).quickSort();
+```
+
+or if you don't want the type argument to the extension to be inferred:
+
+```dart
+MyList<String>(object).quickSort();
+```
+
+or if you imported the extension with a prefix to avoid name collision:
+
+```dart
+prefix.MyList<String>(object).quickSort();
+```
+
+The syntax looks like a constructor invocation, but it does not create a new object.
+
+If `object.quickSort()` would invoke an extension method of `MyList`, then `MyList(object).quickSort()` will invoke the exact same method in the same way.
+
+This mode of invocation also allows invocation of extensions that are *not* in scope, perhaps as `prefix.MyExtension(object).extensionMember()` if imported with a prefix.
+
+The syntax is not *convenient*&mdash;you have to put the "constructor" invocation up front, which removes the one advantage that extension methods have over normal static methods. It is not intended as the common use-case, but as an escape hatch out of conflicts.
+
+### Static Members and Member Resolution
+
+Static member declarations in the extension declaration can be accessed the same way as static members of a class or mixin declaration: By prefixing with the extension's name.
+
+Example:
+
+```dart
+extension MySmart on Object {
+  smart() => smartHelper(this);  // valid
+  static smartHelper(Object o) { ... }
+}
+...
+  MySmart.smartHelper(someObject);  // valid
+```
+
+### Semantics of Invocations
+
+If an extension is found to be the one applying to a member invocation, then at run-time, the invocation will perform a method invocation of the corresponding instance member of the extension, with `this` bound to the receiver value and type parameters bound to the types found by static inference.
+
+If the receiver is `null`, then that invocation is an immediate run-time error unless the `on` type of the extension has a trailing `?`.
+
+With NNBD types, it is a run-time error if the receiver is `null` and the instantiated `on` type of the selected extension does not allow `null`. For sound non-nullable types, this can be excluded statically. If a receiver expression is nullable, then the matching `on` type must be nullable too. If a receiver expression is soundly non-nullable, then the value cannot be `null` at run-time. During migration, we will have unsound nullable types like `int*` which should be matched by `extension Wot on int {…}`, and in that case we will need a run-time check to avoid passing `null` to that extension.
```

Comment:
This paragraph is ordered a little funny. The first sentence makes it sound like there's no static checking of nullable receivers in NNBD but then you get to the end and it seems that's only related to legacy types? It might be good to reorder this to be clearer.